presto-main/src/test/java/com/facebook/presto/spiller/TestSpillCipher.java

```diff
@@ -0,0 +1,80 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spiller;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import org.testng.annotations.Test;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertNotEquals;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.expectThrows;
+
+public class TestSpillCipher
+{
+    @Test
+    public void testAESSpillCipher()
+    {
+        testSpillCipherInstance(new AesSpillCipher());
+    }
+
+    static void testSpillCipherInstance(SpillCipher spillCipher)
+    {
+        testSpillCipherRoundTrips(spillCipher);
+        // Doesn't reuse IV values
+        assertNotEquals(spillCipher.encrypt(new byte[0]), spillCipher.encrypt(new byte[0]), "IV values must not be reused");
+        // Implements cleanup correctly
+        testSpillCipherCleanup(spillCipher);
+    }
+
+    private static void testSpillCipherRoundTrips(SpillCipher spillCipher)
+    {
+        // 0 - 257 byte sizes should ensure that padding cases are all checked
+        for (int size = 0; size <= 257; size++) {
+            testRoundTripData(spillCipher, randomBytes(size), "round-trip data with size: " + size);
+        }
+    }
+
+    private static void testSpillCipherCleanup(SpillCipher spillCipher)
+    {
+        assertFalse(spillCipher.isDestroyed(), "must not be destroyed before check");
+        byte[] encrypted = spillCipher.encrypt(randomBytes(1));
+        spillCipher.destroy();
+        assertTrue(spillCipher.isDestroyed(), "must be destroyed after destroy()");
+        spillCipher.destroy(); // should not throw an exception
+        PrestoException decryptError = expectThrows(PrestoException.class, () -> spillCipher.decrypt(encrypted));
+        assertEquals(decryptError.getErrorCode(), StandardErrorCode.GENERIC_INTERNAL_ERROR.toErrorCode());
+        PrestoException encryptError = expectThrows(PrestoException.class, () -> spillCipher.encrypt(randomBytes(1)));
+        assertEquals(encryptError.getErrorCode(), StandardErrorCode.GENERIC_INTERNAL_ERROR.toErrorCode());
+    }
+
+    private static void testRoundTripData(SpillCipher spillCipher, byte[] data, String message)
+    {
+        assertEquals(data, spillCipher.decrypt(spillCipher.encrypt(data.clone())), message);
```

Comment:
why `data.clone()` and not just `data`?

Reply:
This prevents the implementations from modifying the contents of the `byte[]` that we then use to `assertEquals` on. Slightly defensive, but preserves the correctness of the assertion.